README.txt
README

Este contador utiliza una tabla mysql para registrar las visitas de cada IP por fecha,
de tal manera que si la IP ya ha visitado la web y han transcurrido más de 24h entonces
se incrementa en uno el número de visitas. 

La cuenta total del número de visitas se produce al sumar todas las IP's de la tabla, 
procedimiento muy rápido porque se realiza con un SELECT count(cuenta).

En la próxima revisión quiero incorporar la gestión de almacenamiento de visitas por fechas
(y horas) de tal manera que el webmaster pueda tener una estimación de las visitas por
tramos. Quedará de parte de ese webmaster presentar la información según sus necesidades.

Se prevé guardar una cookie con el valor de las visitas a la web, de tal manera que en la 
próxima visita se vea la animación desde el número de visitas de la última vez hasta las que
hubiera ahora.

De momento no son parametrizables en la llamada algunos valores que luego si se pretender
configurar.

Joseluis Laso 

Probado con los siguientes navegadores

en W-XP: IE8, Safari 5.1.2, FF 9.0.1, Chrome 16.0

si encuentras algún navegador en el que test.php no funcione correctamente por favor, hazme
saber navegador, versión y S.O. para poder solventarlo, gracias.


CHANGELOG.txt
01/02/2012
----------

- Se incluye control de cookies en un ejemplo php para poder controlar el valor
    del contador la última vez que se visitó la página y animar el contador hasta
    el número de visitas que haya actualmente, dando más vistosidad al contador.



test.php
<?php

/*
 * Probado con IE8, Safari 5.1.2, FF 9.0.1, Chrome 16.0 en Win-XP
 */

?>
<html> 
    <head>
     
    <title>Demo de contador | m&aacute;s info en http://jaitec.net/blog/contador-visitas</title>
    <script type="text/javascript" src="firebug-lite/build/firebug-lite.js"></script>
    <!-- thanks to Dean Edwards -->
    <!--[if lt IE 7]>
        <script src="/ie7/ie7.js" type="text/javascript"></script>
    <![endif]-->
    <!--[if IE 8]>
        <script src="/ie7/ie8.js" type="text/javascript"></script>
    <![endif]-->
    <!--[if IE 9]>
        <script src="/ie7/ie9.js" type="text/javascript"></script>
    <![endif]-->    
    <?php require(dirname(__FILE__).'/contador.inc.php'); ?>


    </head>
    
    <body>
        
    <div>
        Contador:                        
        <div id ="contador"></div>
    </div>
  </body>
    <script type="text/javascript">
       window.onload = function(){
           var contador = new Contador(0,999999,6,1,"contador","/contador",42,17);
           contador.show();
           contador.animar();
       }
    </script>
</html>


testphp.php
<?php

/*
 * Probado con IE8, Safari 5.1.2, FF 9.0.1, Chrome 16.0 en Win-XP
 */

// este archivo es un ejemplo de una situación real

include "h.php";    // esto establece las cokies para animar el contador
                    // el visitante ve que el contador pasa del último numero 
                    // de visitas cuando estuvo a las actuales

?>
<html> 
    <head>
     
    <title>Demo de contador | m&aacute;s info en http://jaitec.net/blog/contador-visitas</title>
    <!--script type="text/javascript" src="firebug-lite/build/firebug-lite.js"></script-->
    <!-- thanks to Dean Edwards -->
    <!--[if lt IE 7]>
        <script src="/ie7/ie7.js" type="text/javascript"></script>
    <![endif]-->
    <!--[if IE 8]>
        <script src="/ie7/ie8.js" type="text/javascript"></script>
    <![endif]-->
    <!--[if IE 9]>
        <script src="/ie7/ie9.js" type="text/javascript"></script>
    <![endif]-->    

    </head>
    
    <body>
        
    <div>
        Contador:                        
        <div id ="contador"></div>
    </div>
    <?php print $contador->showAnim('contador');  ?>    
  </body>
</html>


contador.php
<?php

/*
 * estructura de la tabla que alberga el contador
  
  CREATE TABLE  `contador` (
     `id` INT NOT NULL AUTO_INCREMENT PRIMARY KEY ,
     `ip` CHAR( 20 ) CHARACTER SET ASCII COLLATE ascii_general_ci NOT NULL ,
     `fecha` TIMESTAMP NOT NULL ,
     `cuenta` INT NOT NULL ,
      UNIQUE (`ip`)
  ) ENGINE = INNODB;
  
 */

@define ('ERR_ORM_NO_ACTIVO','ORM no está activo');

class Contador {
    
   private $orm = null;
   private $error = '';
   // segundos que han de transcurrir para que se considere una nueva visita
   private $lapso = 86400; // 24h * 60m * 60s;
   // path de las imagenes, relativas al index
   private $pathImg = '.';
   // total acumulado, para no tener que hacer la consulta cada vez
   private $total = -1;
   // total anterior a la actualización para poder hacer animiación del cambio
   private $totalAnt = -1;
 
   public function __construct(PDO $orm, $lapso = 86400, $pathImg='.'){
       $this->lapso = $lapso;
       $this->pathImg = $pathImg;
       if ($orm){
           $this->orm = $orm;
           $this->totalAnt = $this->total($orm);
           $this->inc($orm);
           $this->total = $this->total($orm);
           return 1;
       }else{
           $this->error = ERR_ORM_NO_ACTIVO;
           return 0;
       }
   }
   
   /* si no se integra con Slim / IdiOrm, hacer esto antes de llamar a contador
    *       
    * $orm = new mysqli($dbhost,$dbuser,$dbpass,$dbdb);
    *
    */
   
   public function getError(){ return $this->error; }
   public function getTotal(){ return $this->total; }
   
   // para establecer el totalAnt en el caso de controlar cookies para el navegador del usuario
   public function setTotalAnt($ant) { $this->totalAnt = $ant; }
   
   public function __destruct() {
       //if ($this->orm) $orm->close();
   }
   
   public function clientIp(){
        if (isset($_SERVER['HTTP_X_FORWARDER_FOR']) && $_SERVER['HTTP_X_FORWARDER_FOR'] != ''){
           $ip = (!empty($_SERVER['REMOTE_ADDR']))?$_SERVER['REMOTE_ADDR']:
            ((!empty($_ENV['REMOTE_ADDR']))?$_ENV['REMOTE_ADDR']:"unknow"); 
           $entries = split('[, ]', $_SERVER['HTTP_X_FORWARDER_FOR']);
           reset($entries);
           while (list(,$entry) = each($entries)){
               $entry = trim($entry);
               if (preg_match("/^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)/",$entry,$ip_list)){
                  $private_ip = array(
                      '/^0\./',
                      '/^127\.0\.0\.1/',
                      '/^192\.168\..*/',
                      '/^172\.((1[6-9])|(2[0-9])|(3[0-1]))\..*/',
                      '/^10\..*/');
                  $found_ip = preg_replace($private_ip, $ip, $ip_list[1]);

                  if ($ip != $found_ip){
                      $ip = $found_ip;
                      break;
                  }
               }
           }
        }else{
           $ip = (!empty($_SERVER['REMOTE_ADDR']))?$_SERVER['REMOTE_ADDR']:
            ((!empty($_ENV['REMOTE_ADDR']))?$_ENV['REMOTE_ADDR']:"unknow");         
        }    
        return $ip;
    }
    
    public function inc(PDO $conn){        
        $id = 0; $cant = 0;
        $ip = self::clientIp();
        $sql = "SELECT id, ip, cuenta,UNIX_TIMESTAMP(fecha) fecha FROM contador WHERE ip = '{$ip}'";
        if ($conn){
            try{
                $result = $conn->query($sql);
                if ($row = $result->fetchObject()){
                    $cant = $row->cuenta+1;
                    if ( date ("U") - $row->fecha > $this->lapso ){
                        $sql = "UPDATE contador SET cuenta = ".$cant." WHERE `id` = {$row->id}";
                        $conn->query($sql);
                    }
                }else{
                    $cant = 1;
                    $sql = "INSERT INTO contador (ip,cuenta) VALUES ('{$ip}',1)";    
                    $conn->query($sql);            
                }
            }catch(Exception $e){
                $this->error = $e->getMessage();
            }
        }
        
    }
    
    public function total(PDO $conn){
        $sql = "SELECT sum(cuenta) total FROM contador";
        if ($conn){
            $result = $conn->query($sql);
            $row = $result->fetchObject();
            return $this->total =  $row->total;
        }else{
            $this->error = ERR_ORM_NO_ACTIVO;
            return $this->total = -1;
        }
    }
    
    public function show(){
        if ($this->total == -1) $this->total();
        $cuenta = sprintf("%06s",$this->total);
        $contador = '';
        for($i=0;$i<strlen($cuenta);$i++) {
            $imagen = substr($cuenta,$i,1);
            $contador .= "&nbsp;<img alt='{$imagen}' src='{$this->pathImg}/{$imagen}.gif'>";
        }
        return $contador;
    }

    public function show2(){
        if ($this->total == -1) $this->total();
        $cuenta = sprintf("%06s",$this->total);
        $contador = '<!--div id="contadorJlaso" style="background-image:url(\''.$this->pathImg.'/fondo.png\');background-repeat:no-repeat;"-->
                        <style type="text/css">
                            .digitJlaso {
                                background-repeat:no-repeat;
                                width:30px;
                                height:45px;
                                _position:absolute;
                                vert-align:middle;
                                padding-top:30px;
                            }
                        </style>';
        for($i=strlen($cuenta)-1;$i>=0;$i--) {
            $imagen = substr($cuenta,-$i-1,1);
            $posic  = -42 *  (ord($imagen) - ord("0"));
            $contador .= "<span alt='{$imagen}' id='digitJlaso{$i}' class='digitJlaso' style=\"background-image:url('{$this->pathImg}/digitos.png');background-position:0 {$posic}px;\"><img src='{$this->pathImg}/transp30x45.png'/></span>";
            }
        $contador .= '<!--/div-->
            <!--script type="text/javascript">
                var dig = document.getElementById("digitJlaso0");
                //alert(dig.style.backgroundPosition);
                var actual = 42, stop = 84;
                function animarDigitJlaso(){
                    dig.style.backgroundPosition="0 -"+actual+"px";
                    //alert(dig.style.backgroundPosition);
                    actual++;       
                    if (actual==stop) clearInterval(intJlaso);
                }
                var intJlaso = setInterval("animarDigitJlaso()", 150);
            </script-->';
        return $contador;
    }
    public function showAnim($midiv){
        $html = file_get_contents($this->pathImg.'/contador.inc.php');
        $html = str_replace("'digitos.png'", "'".$this->pathImg."/digitos.png'", $html);
        $html .= '<script type="text/javascript">
                       window.onload = function(){
                           var contador = new Contador('.$this->totalAnt.','.
                                                         $this->total.',
                                                         6, 100, "'.$midiv.'", "'.
                                                         $this->pathImg.'",42,17);
                           contador.show();
                           contador.animar();
                       }            
                  </script>';
        return $html;
    }
    public function incAndGetTotAsImage(PDO $conn,$dir='inc/contador'){
        self::inc($conn);
        $total = self::total($conn);
        return self::show($total,$dir);
    }
    
}

?>


contador.inc.php
<style type="text/css">
#contador{
    display:inline;
}
.digitJlaso {
    background-image:url('digitos.png');
    background-position: 0px 0px;
    background-repeat:no-repeat;
    width:30px;
    height:45px;
    vert-align:middle;
    padding-top:30px;
    *padding-top: 0px;  /* para que solo lo tome en cuenta IE7 e inferiores */
}
</style>
<script type="text/javascript">
    
    /*
     * Objeto contador
     * 
     */
    function Contador(
            desde,      // desde que número vamos a contar
            hasta,      // hasta que número vamos a contar
            digitos,    // digitos de resolución del contador (pone ceros delante)
            milisegs,   // milisegs que transcurren entre cada iteración de animación
            miDiv,      // miDiv, ID del div donde hay que presentar el contador
            pathImg,    // ruta de las imágenes 
            altoImg,    // alto de cada dígito dentro de la tira de imágenes
            offsetImg   // offset del cero dentro de la tira de imágenes
        ){
        this.stop     = hasta;
        this.digitos  = digitos;
        this.milisegs = milisegs;
        this.miDiv    = miDiv;
        this.pathImg  = pathImg;
        this.altoImg  = altoImg;
        this.offsetImg= offsetImg;
        this.actualP  = new Array();
        this.stopP    = 0; 
        this.actual   = desde.toString();
        for (var i=this.actual.length;i<this.digitos;i++){
            this.actual = "0"+this.actual;
        }  
    }

    Contador.prototype = {
        show: function(){
            var a = this.actual.split("");
            var contenedor = document.getElementById(this.miDiv);
            //contenedor.innerHTML += this.actual + "<br/>";
            var e = this.digitos-1;
            var valor=0;
            for (i=0;i<a.length;i++){
                valor = parseInt(a[i]) * this.altoImg + this.offsetImg;
                contenedor.innerHTML += //a[i]+
                  '<a href="http://jaitec.net/blog/contador-visitas" target="_blank">'+
                    '<span id="dig'+e+'" class="digitJlaso" style="background-position: 0px -'+valor+'px;">'+
                    '    <img src="'+this.pathImg+'/transp30x45.png"/>'+
                    '</span>'+
                  '</a>';
                e--;
            }                        
        },
        animarDigito:function(){
            this.actualP[0]++; 
            var dig = document.getElementById("dig0");
            dig.style.backgroundPosition="0 -"+this.actualP[0]+"px"; 
            // mirar si es un nueve para incrementar el de al lado tambien
            d=0;
            while(this.actualP[d]>9*this.altoImg+this.offsetImg){
                d++;
                this.actualP[d]++; 
                var dig = document.getElementById("dig"+d);
                dig.style.backgroundPosition="0 -"+this.actualP[d]+"px"; 
            }
            if (this.actualP[0]>this.stopP){                
                clearInterval(this.interval);
                if (this.actual<this.stop-1){
                    this.actual++;
                    this.animar();
                }
            }
        },
        animar: function(){   
            if (this.actual<this.stop){ 
                var a = this.actual.toString().split("").reverse();
                for (i=0;i<a.length;i++){
                    this.actualP[i] = this.altoImg * a[i] + this.offsetImg;                        
                }
                this.stopP = this.actualP[0] + this.altoImg; 
                var that = this;
                this.interval = setInterval(function(){that.animarDigito();},this.milisegs);
            }
        }
    }
    
    /* modelo de llamada
    window.onload = function(){
       var contador = new Contador(0,999999,6,100,"contenedor","inc/contador",42,17);
       contador.show();
       contador.animar();
    }
    */
</script>